File: day13/day13.go
package main

import (
	//utils "adventofcode2021"
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

type Paper struct {
	dots map[Coordinate]struct{}
	maxX int
	maxY int
}

type Coordinate struct{
	x int
	y int
}

func (p *Paper) fillPaper(rawInput string) {
	for _, c := range strings.Split(rawInput, "\n") {
		splitCoordinate := strings.Split(c, ",")
		x, _ := strconv.Atoi(splitCoordinate[0])
		y, _ := strconv.Atoi(splitCoordinate[1])
		p.dots[Coordinate{x, y}] = struct{}{}
		if x > p.maxX {
			p.maxX = x
		}
		if y > p.maxY {
			p.maxY = y
		}
	}
}

func (p *Paper) print() {
	var paperGrid [][]string
	for i:=0; i < p.maxY+1; i++ {
		col := strings.Split(strings.Repeat(".", p.maxX+1), "")
		paperGrid = append(paperGrid, col)
	}
	for c := range p.dots {
		paperGrid[c.y][c.x] = "#"
	}
	for _, row := range paperGrid {
		fmt.Println(row)
	}
	fmt.Println()
}

func (p *Paper) fold(instruction string) {
	//fmt.Println(instruction)
	re := regexp.MustCompile(`fold along (x|y)=([0-9]+)`)
	parsedInstructions := re.FindStringSubmatch(instruction)
	direction := parsedInstructions[1]
	lineNum, _ := strconv.Atoi(parsedInstructions[2])

	switch direction {
	case "x": {
		for c := range p.dots {
			if c.x > lineNum {
				delete(p.dots, c)
				c.x = lineNum - (c.x-lineNum)
				p.dots[c] = struct{}{}
			}

		}
		p.maxX = lineNum-1
	}
	case "y": {
		for c := range p.dots {
			if c.y > lineNum {
				delete(p.dots, c)
				c.y = lineNum - (c.y-lineNum)
				p.dots[c] = struct{}{}
			}
		}
		p.maxY = lineNum-1
	}
	}
	//p.print()
	//fmt.Println(p.dots)
	//fmt.Println(len(p.dots))
	//fmt.Println()
}

func main() {
	contents, _ := ioutil.ReadFile("inputs/day13.txt")
	sliceContents := strings.Split(string(contents), "\n\n")
	paperInput := sliceContents[0]
	folding := sliceContents[1]

	paperGraphMap := make(map[Coordinate]struct{})
	paperGraph := Paper{paperGraphMap, 0, 0}
	paperGraph.fillPaper(paperInput)
	//fmt.Println(paperGraph.dots)
	//paperGraph.print()

	for _, fold := range strings.Split(folding, "\n"){
		paperGraph.fold(fold)
	}
	paperGraph.print()
}

File: day13/day13.py
def print_paper(paper, max_x, max_y, fold_dir="", fold_line=-1):
    paper_grid = []
    for i in range(max_y+1):
        paper_grid.append(["." for _ in range(max_x+1)])

    for x, y in paper:
        paper_grid[y][x] = "#"

    if fold_dir == "x":
        for g in paper_grid:
            g[fold_line] = "|"
    if fold_dir == "y":
        paper_grid[fold_line] = ["-" for _ in range(max_x+1)]

    for g in paper_grid:
        print("".join(g))
    print()


def fold_over(paper, fold_dir, fold_line):


if __name__ == "__main__":
    with open("../tests/day13.txt") as file:
        lines = [x.strip() for x in file.readlines()]
    i = 0
    max_x = 0
    max_y = 0
    positions = set()
    for input_spot in lines:
        if input_spot == '':
            break
        input_spot = input_spot.split(",")
        x = int(input_spot[0])
        y = int(input_spot[1])
        positions.add((x, y))
        max_x = max(x, max_x)
        max_y = max(y, max_y)
        i += 1

    instructions = []
    for c in range(i+1, len(lines)):
        instruction = lines[c].split(' ')[-1]
        instructions.append(instruction)

    print(positions)
    print(instructions)

    for fold in instructions:
        fold = fold.split("=")
        direction = fold[0]
        line_num = int(fold[1])
